File: src/main/java/city/newnan/api/i18n/LanguageManager.java
package city.newnan.api.i18n;

import city.newnan.api.config.ConfigManager;
import city.newnan.api.message.MessageManager;
import me.lucko.helper.terminable.Terminable;
import org.bukkit.plugin.Plugin;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class LanguageManager implements MessageManager.LanguageProvider, Terminable {

    private Language majorLanguage;
    private Language fallbackLanguage;
    private final Map<Locale, Language> languageMap = new HashMap<>();
    private final Plugin plugin;

    public LanguageManager(Plugin plugin) {
        this.plugin = plugin;
    }

    public LanguageManager register(Locale locale, String filePath) throws FileNotFoundException, IOException, ConfigManager.UnknownConfigFileFormatException {
        File file = new File(plugin.getDataFolder(), filePath);
        if (file.exists() || file.isFile()) {
            Language language = new Language(file, locale);
            languageMap.put(locale, language);
            if (majorLanguage == null) {
                majorLanguage = language;
            }
            else if (fallbackLanguage == null) {
                fallbackLanguage = language;
            }
        } else {
            throw new FileNotFoundException(filePath);
        }
        return this;
    }

    public LanguageManager unregister(Locale locale) {
        languageMap.remove(locale);
        return this;
    }

    public Language getLanguage(Locale locale) {
        return languageMap.get(locale);
    }

    public LanguageManager setMajorLanguage(Locale locale) {
        Language language = languageMap.get(locale);
        if (language != null) {
            majorLanguage = language;
        }
        return this;
    }

    public LanguageManager setFallbackLanguage(Locale locale) {
        Language language = languageMap.get(locale);
        if (language != null) {
            fallbackLanguage = language;
        }
        return this;
    }

    public LanguageManager reloadAll() {
        languageMap.forEach((locale, language) -> {
            try {
                language.reload();
            } catch (IOException | ConfigManager.UnknownConfigFileFormatException e) {
                e.printStackTrace();
            }
        });
        return this;
    }

    private static final Pattern pattern = Pattern.compile("[$]([^$]+?)[$]");

    @Override
    public String provideLanguage(String rawText) {
        Matcher matcher = pattern.matcher(rawText);
        while (matcher.find()) {
            String key = matcher.group(0);
            String path = key.replace("$", "");
            String replaced;
            if (majorLanguage == null || (replaced = majorLanguage.getNodeString(path)) == null) {
                if (fallbackLanguage == null || (replaced = fallbackLanguage.getNodeString(path)) == null)
                    continue;
            }
            rawText = rawText.replace(key, replaced);
        }

        return rawText.replaceAll("[$][$]", "$");
    }

    @Override
    public void close() {
        languageMap.clear();
    }

    public static class FileNotFoundException extends Exception {
        public FileNotFoundException(String path) {
            super("File not found: " + path);
        }
    }
}

File: src/main/kotlin/city/newnan/violet/i18n/LanguageManager.kt
package city.newnan.violet.i18n

import city.newnan.violet.config.ConfigManager.UnknownConfigFileFormatException
import city.newnan.violet.message.LanguageProvider
import me.lucko.helper.terminable.Terminable
import org.bukkit.plugin.Plugin
import java.io.File
import java.io.IOException
import java.util.*
import java.util.regex.Pattern

private val pattern = Pattern.compile("[$][^$]+[$]")

class LanguageManager(private val plugin: Plugin) : LanguageProvider, Terminable {
    var majorLanguage: Language? = null
        private set
    var defaultLanguage: Language? = null
        private set
    private val languageMap: MutableMap<Locale, Language> = HashMap()

    @Throws(FileNotFoundException::class, IOException::class, UnknownConfigFileFormatException::class)
    fun register(locale: Locale, filePath: String): LanguageManager {
        val file = File(plugin.dataFolder, filePath)
        if (file.exists() || file.isFile) {
            languageMap[locale] = Language(file, locale)
        } else {
            throw FileNotFoundException(filePath)
        }
        return this
    }

    infix fun unregister(locale: Locale): LanguageManager = this.also { languageMap.remove(locale) }

    infix fun getLanguage(locale: Locale): Language? = languageMap[locale]

    infix fun setMajorLanguage(locale: Locale): LanguageManager = this.also { languageMap[locale]?.run { majorLanguage = this } }

    fun guessMajorLanguage() {
        Locale.getDefault().run {
            if (languageMap.containsKey(this)) {
                setMajorLanguage(this)
            }
        }
    }

    infix fun setDefaultLanguage(locale: Locale): LanguageManager = this.also { languageMap[locale]?.run {defaultLanguage = this  } }

    fun reloadAll(): LanguageManager {
        languageMap.forEach { (locale: Locale?, language: Language) ->
            try {
                language.reload()
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: UnknownConfigFileFormatException) {
                e.printStackTrace()
            }
        }
        return this
    }

    override infix fun provideLanguage(rawText: String): String {
        var rawTextShadow = rawText
        val matcher = pattern.matcher(rawTextShadow)
        while (matcher.find()) {
            val key = matcher.group(0)
            val path = key.replace("$", "")
            var replaced: String? = null
            if (majorLanguage == null || majorLanguage!!.getNodeString(path).also { replaced = it } == null) {
                if (defaultLanguage == null || defaultLanguage!!.getNodeString(path)
                        .also { replaced = it } == null
                ) continue
            }
            rawTextShadow = rawTextShadow.replace(key, replaced!!)
        }
        return rawTextShadow.replace("[$][$]".toRegex(), "$")
    }

    override fun close() = languageMap.clear()

    class FileNotFoundException(path: String) : Exception("File not found: $path")
}